File: Python3/프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    # 최소 시간은 1분
    left = 1
    # 최대 시간은 가장 느린 검사관이 모든 사람을 검사할 때 걸리는 시간
    right = max(times) * n
    
    # 이진 탐색 시작
    while left < right:
        mid = (left + right) // 2  # 중간 값을 계산
        # 모든 검사관이 중간 값 시간 내에 검사할 수 있는 사람 수의 합
        total = sum(mid // time for time in times)  
        
        if total >= n:
            # 검사할 수 있는 사람 수가 충분하면 시간을 줄임
            right = mid
        else:
            # 검사할 수 있는 사람 수가 부족하면 시간을 늘림
            left = mid + 1
    
    # 최소 시간이 left에 저장됨
    return left

File: C++/프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <vector>
#include <algorithm>

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    long long left = 1;
    long long right = *max_element(times.begin(), times.end()) * (long long)n;

    while (left <= right) {
        long long mid = (left + right) / 2;
        long long total = 0;
        
        // mid 시간 동안 각 심사관이 몇 명을 처리할 수 있는지 계산
        for (int time : times) {
            total += mid / time;
        }

        if (total >= n) {
            answer = mid;  // 더 짧은 시간으로도 가능할 수 있으므로 답 갱신
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    return answer;
}
